universe/ValueRefs.h

```diff
@@ -377,6 +417,19 @@ struct FO_COMMON_API Operation final : public ValueRef<T>
 
     unsigned int GetCheckSum() const override;
 
+    std::unique_ptr<ValueRef<T>> Clone() const override {
+        std::vector<std::unique_ptr<Operation<T>>> operands;
+        operands.reserve(m_operands.size());
+        for (const std::unique_ptr<ValueRef<T>>& operand : m_operands) {
+            if (operand) {
+                operands.push_back(std::move(operand->Clone()));
+            } else {
+                operands.emplace_back(nullptr);
+            }
+        }
+        return std::make_unique<Operation<T>>(m_op_type, std::move(operands));
```

Comment:
Would perhaps be nice to clone all the invariants directly rather than relying on the constructor re-determining them.

Reply:
Could consructor determine them differently?

Reply:
If you construct a new ValueRef / Condition using the standard constructors, it just gets pointers to sub-refs and sub-condtions and needs to check them all to determine its own invariance.

You could also set up a base class virtual `Copy` method and use that in `Clone` as suggested here: https://github.com/freeorion/freeorion/pull/3309#issuecomment-791911615 and copy the invariants in the base class implementations.

I don't think it makes sense to also set up constructor overloads that take invariants and pass them to the base classes similar overloads.